File: PS6/readability.py
from cs50 import get_string
import re


def main():
    text = get_string("Text: ")
    
    # Computes for index from Coleman-Liau formula
    L = (count_letters(text) / count_words(text)) * 100
    S = (count_sen(text) / count_words(text)) * 100
    index = 0.0588 * L - 0.296 * S - 15.8
    
    if (index > 16):
        print("Grade 16+")
    elif (index < 1):
        print("Before Grade 1")
    else:
        print(f"Grade {int(round(index))}")
        

# Counts number of words using split function
# Split function separates text per whitespace character
def count_words(text):
    count = 0
    for word in text.split():
        count += 1
    return count
    

# Counts number of letters in each word
def count_letters(text):
    count = 0
    # Searches for alphanumeric set of characters separated by non-alphanumeric characters
    for w in re.findall('\w+', text):
        for c in w:
            count += 1
    return count
    

# Counts number of punctuations (period, exclamation point, question mark)
def count_sen(text):
    count = 0
    for punc in re.findall(r'[.!?]', text):
        count += 1
    return count
    
    
main()

File: PS2/readability.c
#include <stdio.h>
#include <cs50.h>
#include <ctype.h>
#include <string.h>
#include <math.h>

// A program that tests readability of a text through the Coleman-Liau index
// index = 0.0588 * L - 0.296 * S - 15.8
// where L is the average # of letters per 100 words
// and S is the average # of sentences per 100 words

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    string text = get_string("Text: "); // Gets text input from user

    // Stores counted letters, words, and sentences into variables
    int l = count_letters(text), w = count_words(text) + 1, s  = count_sentences(text);

    // L = ave # of letters per 100 words, S = ave # of sentences per 100 words
    float L = (l / (float)w) * 100, S = (s / (float)w) * 100;

    // Coleman-Liau index formula
    float index = 0.0588 * L - 0.296 * S - 15.8;

    if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", (int)round(index)); // Rounds up computed index and prints as integer
    }

}

// Counts number of letters in text
int count_letters(string text)
{

    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++) // Loops through each char in text
    {
        if (isalpha(text[i]) != 0) // Checks if char is an alphabetic char
        {
            count++; // Counts every alphabetic char found
        }
    }
    return count; // Returns number of letters in text
}

// Counts number of words in text
int count_words(string text)
{

    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++) // Loops through each char in text
    {
        if (isspace(text[i]) != 0) // Checks if char is a white-space char
        {
            count++; // Counts every white-space char found
        }
    }
    return count; // Returns number of spaces in text
}

// Counts number of sentences in text
int count_sentences(string text)
{

    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++) // Loops through each char in text
    {
        if (text[i] == '.' || text[i] == '?' || text[i] == '!') // Checks if char is a '.', '?', or '!'
        {
            count++; // Counts every '.', '?' '!' char found
        }
    }
    return count; // Returns number of punctuations in text
}